hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/UnassignRegionHandler.java

```diff
@@ -76,75 +79,75 @@ private HRegionServer getServer() {
-      // XXX: The behavior is a bit broken. At master side there is no FAILED_CLOSE state, so if
-      // there are exception thrown from the CP, we can not report the error to master, and if here
-      // we just return without calling reportRegionStateTransition, the TRSP at master side will
-      // hang there for ever. So here if the CP throws an exception out, the only way is to abort
-      // the RS...
-      region.getCoprocessorHost().preClose(abort);
-    }
-    if (region.close(abort) == null) {
-      // XXX: Is this still possible? The old comment says about split, but now split is done at
-      // master side, so...
-      LOG.warn("Can't close region {}, was already closed during close()", regionName);
-      rs.getRegionsInTransitionInRS().remove(encodedNameBytes, Boolean.FALSE);
-      return;
-    }
-    rs.removeRegion(region, destination);
-    if (!rs.reportRegionStateTransition(new RegionStateTransitionContext(TransitionCode.CLOSED,
-      HConstants.NO_SEQNUM, closeProcId, -1, region.getRegionInfo()))) {
-      throw new IOException("Failed to report close to master: " + regionName);
+      HRegion region = rss.getRegion(this.encodedRegionName);
+      // Can't unassign a Region that is not online.
+      if (region == null) {
+        LOG.debug("Received CLOSE for {} which is not online, and we're not opening/closing.",
+          this.encodedRegionName);
+        // Remove what we added above.
+        rss.getRegionsInTransitionInRS().remove(encodedRegionNameAsBytes, Boolean.FALSE);
+        return;
+      }
+      CloseRegionHandler crh = new CloseRegionHandler(rss, region.getRegionInfo(), this.abort,
+          this.destination) {
+        @Override boolean reportRegionStateTransition() throws IOException {
+          boolean b = super.reportRegionStateTransition();
+          if (!b) {
+            String name = getRegionInfo().getRegionNameAsString();
+            throw new IOException("Failed report close of " + name + " to master");
+          }
+          return b;
+        }
+
+        @Override long getClosePID() {
+          return UnassignRegionHandler.this.closeProcId;
+        }
+      };
+      crh.process();
+    } finally {
+      // Cache the close region procedure id
+      rss.finishRegionProcedure(closeProcId);
     }
-    // Cache the close region procedure id after report region transition succeed.
-    rs.finishRegionProcedure(closeProcId);
-    rs.getRegionsInTransitionInRS().remove(encodedNameBytes, Boolean.FALSE);
-    LOG.info("Closed {}", regionName);
   }
 
   @Override
   protected void handleException(Throwable t) {
-    LOG.warn("Fatal error occurred while closing region {}, aborting...", encodedName, t);
-    getServer().abort("Failed to close region " + encodedName + " and can not recover", t);
+    LOG.warn("Fatal error occurred while closing {}, aborting...", this.encodedRegionName, t);
+    getServer().abort("Failed to close " + this.encodedRegionName, t);
```

Comment:
Please add the 'can not recover' back? It clearly tells why we must abort the region server.

Reply:
Its aborting so seems plain to me that 'it can not recover' but no problem, will add it back.